Major enhancements to ShadowMediaPlayer

kriegfrj opened the pull request:
This PR represents the final culmination of many months' part-time work after finally managing to back out of the several rabbit warrens that I was forced to go down. :wink:

Implementing media playback in Android (and probably in general) is tricky because of all the different asynchronous events and errors that can happen, and the requirements to make sure you are calling the right methods in the right states. This shadow implementation attempts to emulate all of the state transitions/error callbacks/exceptions that Android does so that you can thoroughly unit test your media playback routines, by (eg) asserting if you call a method while in the wrong state. Much of this was gleaned from the `MediaPlayer` documentation, but because the documentation was sometimes difficult to understand I also wrote a little test program & ran it on my Android device to compare documented behaviour with actual behaviour. Unsurprisingly, the two were not always the same; where they differed I implemented whichever of the two were stricter in order to encourage the most robust possible developer code.

This implementation also allows you to schedule playback events (errors, buffer underruns, info messages, etc) to occur during simulated playback, and you can control playback progess via the calling thread Looper's Scheduler instance (including pausing, advancing manually, or allowing the whole playback to run all the way to completion).

There is some Javadoc that explains the main features.

It also happens to be backward compatible with the old `ShadowMediaPlayer` - however, with the big changes in the works for Robolectric 3.0 this is probably not as big a selling point as it was when I started. Arguably now the "SILENT" error state emulation mode should not be the default (and arguably could be removed completely).

Comments/criticisms would be very welcome.

erd commented:
As someone who's built a media player app before, this is pretty awesome :+1: 

There are a bunch of commented out tests in `ShadowMediaPlayerTest`. Are these things that you plan on fixing? If not, I'd prefer to simply delete the tests.

kriegfrj commented:
Glad you like it Erich! Btw, do you have an opinion on the invalid state behaviour, and what should be the default?

I have re-instated some of those commented tests and removed the rest. One of those removed was redundant; the others were for testing changes to MediaInfo during playback (which was supported in an earlier revision but I dropped support when I made the media info hash static).

It also occurred to me that setDataSource() (or later in the code) will fail with a NPE if you set it do a data source that hasn't either got an exception or `MediaInfo` instance associated with it. This makes it not backward compatible, and it also can cause a delayed failure. I'm trying to think of ways to fix this but if you're happy with what is already here I can do this as a future PR.

kriegfrj commented:
> It also occurred to me that setDataSource() (or later in the code) will fail with a NPE if you set it do a
> data source that hasn't either got an exception or MediaInfo instance associated with it. This makes 
> it not backward compatible, and it also can cause a delayed failure. I'm trying to think of ways to fix 
> this but if you're happy with what is already here I can do this as a future PR.

Ok, so I decided to go ahead and implement a fix for this, but I'm happy to take suggestions on whether it is the best way to proceed or if we should take a different approach.

Per the commit message, I changed it so that it asserts if you call `setDataSource()` and neither an exception nor a `MediaInfo` has been configured for that data source. This produces a fail-fast situation which is (I think) better than what I had before. However, I have the following questions:

1. Should assertions/exceptions differentiate between errors in the Application-Under-Test (AUT) and errors in the test itself? I feel that assertions should generally be used to indicate an error in the AUT, but failing to set a `MediaInfo` would classify as a problem in the test. Should I take some other action rather than asserting?
2. Current implementation requires that the media info be set before `setDataSource()` is called. This complicates test setup for `MediaPlayer`-related tests. Is this extra complication justified? Should we have a default media info (possibly configurable) that is used if no matching media info is found for a given data source? 
3. As a possible future enhancement, we could register a callback with the shadow which is invoked when `setDataSource()` is called if no matching `MediaInfo` is found. This could allow programmatic/dynamic setting of the media info rather than having to register them all in the hash before the test starts.

I guess I would like to see what the real-world use-cases are like (including in my own tests which are coming) before going to this effort, particularly with 2 & 3.  However, seeing as how you said you've had experience developing a media player app I was wondering if you had any thoughts Erich.

erd commented:
> Should assertions/exceptions differentiate between errors in the Application-Under-Test (AUT) and errors in the test itself? I feel that assertions should generally be used to indicate an error in the AUT, but failing to set a MediaInfo would classify as a problem in the test. Should I take some other action rather than asserting?

I don't have a strong feeling here. As long as it fails fast, I'm ok with either approach.

> Current implementation requires that the media info be set before setDataSource() is called. This complicates test setup for MediaPlayer-related tests. Is this extra complication justified? Should we have a default media info (possibly configurable) that is used if no matching media info is found for a given data source?

I think requiring the user to populate the MediaPlayer beforehand is fine. If the user wants a simpler approach to testing a MediaPlayer, you can always use Mockito (this is what I've done in the past).

> However, seeing as how you said you've had experience developing a media player app I was wondering if you had any thoughts Erich.

In the past, I've created a thin wrapper around MediaPlayer and tested it by hand on a device, then mocked the wrapper to assert that it's collaborators were interacting with it correctly. I'd have to try writing some tests with this new emulated shadow to have a stronger opinion on how I'd improve it. I'd be in favor of merging this before doing #3 and see what else is needed for testing.

kriegfrj commented:
> I'd be in favor of merging this before doing #3 and see what else is needed for testing.

Agreed. I've changed the assertion to an IllegalArgumentException per your suggestion and committed this, along with some warnings in the Javadoc. Ready to merge as far as I'm concerned (the error in Travis for JDK1.8 looks to be unrelated, though it could be a legitimate issue that cropped up due to the particular ordering of the tests).
